### PR TITLE
feat: ERC1155 clientside changes

### PR DESCRIPTION
**Pull request by phated**
@cha0sg0d this is some massive hacked-to-pieces work that gets the client "running" again with the artifact changes. Please note that I felt the need to restructure your state for artifacts by attaching them to the relevant planet so I could more easily have all the data I needed when pulling a planet from the contract.

I disabled a TON of stuff just to get the client compiling and launching again, but this is going to be my reference for the stuff we need to get done (hence the Draft).

### CODE REVIEW

**Comment by cha0sg0d** [comment]
Ok there is a lot of stuff that doesn't work, but the contract changes look good to me. 2 big questions / comments:
1. Do we need a way to store artifact / spaceship voyages for easier fetching? Right now they are only fetchable by querying the Arrival that has their tokenId.
2. DecodeArtifact is broken because the indexes for fetching info at specific bytes are all `0`. I assume this is on purpose so this is a reminder that that piece of code needs to be updated 😇.

**Comment by cha0sg0d** [comment]
Screen Shot 2022-10-03 at 10 56 47 AM

Also get this error when actually trying to start up a game.

File: packages/serde/src/spaceship.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * can be used if you need to pass an spaceship ID into a web3 call.
+ *
+ * @param spaceshipId non-0x-prefixed lowercase hex `string` of 64 hex characters
+ * representing an spaceship's ID
+ */
+export function spaceshipIdToDecStr(spaceshipId: SpaceshipId): string {
+  return bigInt(spaceshipId, 16).toString(10);
+}
+
+function calculateByteUInt(tokenId: EthersBN, startByte: number, endByte: number) {
```
Comment (phated) [inline packages/serde/src/spaceship.ts:62]: TODO: A shared utility function?

File: packages/serde/src/artifact.ts [hunk 2]
```diff
@@ -84,37 +88,42 @@ export function decodeArtifactPointValues(
...
-export function decodeArtifact(rawArtifactWithMetadata: RawArtifactWithMetadata): Artifact {
-  const { artifact, owner, upgrade, timeDelayedUpgrade, locationId, voyageId } =
-    rawArtifactWithMetadata;
+export function decodeArtifact(tokenId: EthersBN): Artifact {
+  // 454093553060631688131761916536684786657829610989267258091508080987871379456
+  // Collection: Artifact
+  // Rarity: Common
+  // Type: Colossus
+  // Biome: Forest
+
```
Comment (phated) [inline packages/serde/src/artifact.ts:111]:
```suggestion
```

File: packages/serde/src/artifact.ts [hunk 3]
```diff
@@ -84,37 +88,42 @@ export function decodeArtifactPointValues(
   };
 }
 
-export type RawArtifactWithMetadata = Awaited<ReturnType<DarkForest['getArtifactById']>>;
+function calculateByteUInt(tokenId: EthersBN, startByte: number, endByte: number) {
```
Comment (phated) [inline packages/serde/src/artifact.ts:91]: TODO: A shared utility function?

File: packages/procedural/src/ArtifactProcgen.ts [hunk 4]
```diff
@@ -2,21 +2,106 @@ import { EMPTY_ADDRESS, EMPTY_LOCATION_ID } from '@dfdao/constants';
...
-  artifactNameFromArtifact,
   ArtifactRarity,
   ArtifactType,
   Biome,
+  Spaceship,
+  SpaceshipId,
 } from '@dfdao/types';
 
-const namesById = new Map<ArtifactId, string>();
+// TODO: get this out of here
```
Comment (phated) [inline packages/procedural/src/ArtifactProcgen.ts:12]:
```suggestion
```

File: packages/serde/src/artifact.ts [hunk 5]
```diff
@@ -84,37 +88,52 @@ export function decodeArtifactPointValues(
...
+  // Type: Colossus
+  // Biome: Forest
+
+  // These acocunt for unknown at the 0-th index
+  const tokenIdx = ArtifactInfo.TokenType - 1;
+  const rarityIdx = ArtifactInfo.ArtifactRarity - 1;
+  const typeIdx = ArtifactInfo.ArtifactType - 1;
+  const biomeIdx = ArtifactInfo.Biome - 1;
+
+  const _tokenType = calculateByteUInt(tokenId, tokenIdx, tokenIdx);
```
Comment (phated) [inline packages/serde/src/artifact.ts:118]: What should happen if something tries to decode an invalid token type? I think we should probably throw in the client too. WDYT?

File: packages/serde/src/spaceship.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * Converts the raw Token ID to an `Spaceship` typescript typed object (see @dfdao/types).
+ *
+ * @param tokenId Raw `tokenId` representing an `Spaceship` struct
+ */
+export function decodeSpaceship(tokenId: EthersBN): Spaceship {
+  // These acocunt for unknown at the 0-th index
+  const tokenIdx = SpaceshipInfo.TokenType - 1;
+  const typeIdx = SpaceshipInfo.SpaceshipType - 1;
+
+  const _tokenType = calculateByteUInt(tokenId, tokenIdx, tokenIdx);
```
Comment (phated) [inline packages/serde/src/spaceship.ts:81]: TODO: Handle this

**Comment by cha0sg0d** [comment]
I got this branch running locally and everything works! Super sorry about the `_beforeTokenTransfer` problem.

What else do we need to do before we can a) merge this into my branch and b) merge into main?

**Comment by cha0sg0d** [comment]
```js
Failed to start plugin: Admin Controls - Please review stack trace
 Error: [Package Error] "decoders" no dependency version info found. (Imported by "@dfdao/types").
    at unknown:decoders?from=@dfdao/types:13:7
    ```
    Get this error when starting admin controls as well

**Comment by phated** [comment]
@cha0sg0d this should be good to go—it's even rebased on your latest changes to the 1155 PR. There are going to be some bugs but I expect you to either fix them as found or open issues that I can look at later.

**Pull request merged**